src/GafferScene/Instancer.cpp

```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
 			}
 		}
 
+		void initPrototypes( PrototypeMode &mode, const std::string &index, const std::string &rootsVariable, const StringVectorData *rootsList )
+		{
+			const std::vector<std::string> *rootStrings = nullptr;
+
+			switch( mode )
+			{
+				case IndexedRootsList :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					rootStrings = &rootsList->readable();
+
+					break;
+				}
+				case IndexedRootsVariable :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					const auto *roots = m_primitive->variableData<StringVectorData>( rootsVariable, PrimitiveVariable::Constant );
+					if( !roots )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Constant StringVectorData when using IndexedRootsVariable mode" );
+					}
+
+					rootStrings = &roots->readable();
+					if( rootStrings->empty() )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must specify at least one root location" );
+					}
+
+					break;
+				}
+				case RootPerVertex :
+				{
+					const auto view = m_primitive->variableIndexedView<StringVectorData>( rootsVariable, PrimitiveVariable::Vertex );
+					if( !view )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Vertex StringVectorData when using RootPerVertex mode" );
+					}
+
+					m_indices = view->indices();
+					rootStrings = &view->data();
+					if( rootStrings->empty() )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must specify at least one root location" );
+					}
+
+					break;
+				}
+			}
+
+			m_roots = new ObjectVector;
+			auto &roots = m_roots->members();
+			roots.reserve( rootStrings->size() );
+
+			std::vector<ConstInternedStringVectorDataPtr> inputNames;
+			inputNames.reserve( rootStrings->size() );
+
+			size_t i = 0;
+			ScenePlug::ScenePath path;
+			for( const auto &root : *rootStrings )
+			{
+				ScenePlug::stringToPath( root, path );
+				roots.emplace_back( new InternedStringVectorData( path ) );
```

Comment:
Should we be checking the existence of the path, and throwing if it's invalid? I think we probably should?

Reply:
I thought path exists checks were too expensive. Happy to add it though.

Reply:
I think here we _should_ be OK, because we don't initialise engines all that frequently. I also have a bit of a plan for improving the performance of existence checks in future so I'm not quite as wary as I once was.

Reply:
Addressed in ea93f16bf034a728ca56fab80ee230bb1c97305d

Reply:
I've added one comment on that commit, since I think it has an unnecessary PathScope.

Reply:
Addressed in 09390311f151b02c4ce5972d9119bfd82a490583